pkg/bootstrap/static.go

```diff
@@ -0,0 +1,392 @@
+				Name:           ZipkinCluster,
+				ConnectTimeout: 1 * time.Second,
+				Type:           v2.Cluster_STRICT_DNS,
+				LbPolicy:       v2.Cluster_ROUND_ROBIN,
+				Hosts:          []*core.Address{zipkinAddress},
+			})
+		out.Tracing = &trace.Tracing{
+			Http: &trace.Tracing_Http{
+				Name: util.Zipkin,
+				Config: toStruct(&trace.ZipkinConfig{
+					CollectorCluster:  ZipkinCluster,
+					CollectorEndpoint: "/api/v1/spans",
+				}),
+			},
+		}
+	}
+
+	// telemetry bootstrap should route telemetry directly to the local upstream
+	telemetryCluster := TelemetryCluster
+	if opts.TelemetryUpstream == nil {
+		telemetryAddress, err := toAddress(opts.TelemetryAddress)
+		if err != nil {
+			return nil, err
+		}
+		cluster := v2.Cluster{
+			Name:                 TelemetryCluster,
+			ConnectTimeout:       1 * time.Second,
+			Type:                 v2.Cluster_STRICT_DNS,
+			LbPolicy:             v2.Cluster_ROUND_ROBIN,
+			Hosts:                []*core.Address{telemetryAddress},
+			CircuitBreakers:      circuitBreakers,
+			Http2ProtocolOptions: &core.Http2ProtocolOptions{},
+		}
+		if opts.ControlPlaneAuth {
+			cluster.TlsContext = &auth.UpstreamTlsContext{
+				CommonTlsContext: commonTLSContext([]string{opts.TelemetrySAN}),
+			}
+		}
+		out.StaticResources.Clusters = append(out.StaticResources.Clusters, cluster)
+	} else {
+		telemetryCluster = clusterName(*opts.TelemetryUpstream)
+	}
+
+	out.StaticResources.Listeners = BuildListeners(opts.Upstreams, telemetryCluster, opts.ControlPlaneAuth)
+
+	if err := out.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &out, nil
+}
+
+func commonTLSContext(altNames []string) *auth.CommonTlsContext {
+	return &auth.CommonTlsContext{
+		TlsCertificates: []*auth.TlsCertificate{{
+			CertificateChain: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/cert-chain.pem"},
+			},
+			PrivateKey: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/key.pem"},
+			},
+		}},
+		ValidationContext: &auth.CertificateValidationContext{
+			TrustedCa: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/root-cert.pem"},
+			},
+			VerifySubjectAltName: altNames,
+		},
+	}
+}
+
+func toAddress(address string) (*core.Address, error) {
+	host, portStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil, err
+	}
+	return &core.Address{
+		Address: &core.Address_SocketAddress{
+			SocketAddress: &core.SocketAddress{
+				Address:       host,
+				PortSpecifier: &core.SocketAddress_PortValue{PortValue: uint32(port)},
+			},
+		},
+	}, nil
+}
+
+func toStruct(msg proto.Message) *types.Struct {
+	out, err := util.MessageToStruct(msg)
+	if err != nil {
+		panic(err)
+	}
+	return out
+}
+
+// BuildServiceConfig generates a mixer client config.
+func BuildServiceConfig(service, uid, telemetryCluster string) *mixerclient.HttpClientConfig {
```

Comment:
Same as above: we already have code that produces the mixer client config, it should be extracted out and re-used here.

https://github.com/istio/istio/blob/master/pilot/pkg/networking/plugin/mixer/mixer.go

Reply:
Disagree, @ZackButcher . This is specifically not using pilot for two reasons: 1) the code I need is in massive churn, 2) this is much simpler as it only has to handle istio components.

Reply:
We have to wire in a lot of these Istio components anyway. I'm also not opposed to pulling in Pilot to this top-level dir given this stuff is exactly Pilot's domain.

Yes there's churn, but implement it once and as the code churns the people updating it will update this too. We really really do not want a lot of duplicate code creating lots of Envoy config when we can have it all in a common place. cc @rshriram @costinm for opinions too, but I'm pretty opposed to this not using libs in Pilot.

Reply:
@ZackButcher we do not want to use Pilot for internal sidecars yet. There is no need to have monopoly on envoy config from pilot code. It is completely fine to write several templates that manipulate envoy protos; in fact, we already do that in several places.

cc @louiscryan : I do not see how this needs to be tied-in with pilot. This is just a simple template generator, why do we need to use pilot for this?